GVFS/GVFS.Platform.Mac/MacFileBasedLock.cs

```diff
@@ -0,0 +1,159 @@
+            LockUn = 8  // #define LOCK_UN   8    /* unlock */
+        }
+
+        public override bool TryAcquireLock()
+        {
+            if (this.lockFileDescriptor == InvalidFileDescriptor)
+            {
+                this.lockFileDescriptor = NativeMethods.Open(this.LockPath, OpenCreate | OpenWriteOnly, FileMode644);
+                if (this.lockFileDescriptor == InvalidFileDescriptor)
+                {
+                    int errno = Marshal.GetLastWin32Error();
+                    EventMetadata metadata = this.CreateEventMetadata(errno);
+                    this.Tracer.RelatedWarning(
+                        metadata,
+                        $"{nameof(MacFileBasedLock)}.{nameof(this.TryAcquireLock)}: Failed to open lock file");
+                    
+                    return false;
+                }
+            }
+
+            if (NativeMethods.FLock(this.lockFileDescriptor, (int)(FLockOperations.LockEx | FLockOperations.LockNb)) != 0)
+            {
+                int errno = Marshal.GetLastWin32Error();
+                if (errno != EIntr && errno != EWouldBlock)
+                {
+                    EventMetadata metadata = this.CreateEventMetadata(errno);
+                    this.Tracer.RelatedWarning(
+                        metadata,
+                        $"{nameof(MacFileBasedLock)}.{nameof(this.TryAcquireLock)}: Unexpected error when locking file");
+                }
+
+                return false;
+            }
+
+            if (NativeMethods.FTruncate(this.lockFileDescriptor, 0) == 0)
+            {
+                byte[] signatureBytes = Encoding.UTF8.GetBytes(this.Signature);
+                long bytesWritten = NativeMethods.Write(
+                    this.lockFileDescriptor,
+                    signatureBytes,
+                    Convert.ToUInt64(signatureBytes.Length));
+
+                if (bytesWritten == -1)
+                {
+                    int errno = Marshal.GetLastWin32Error();
+                    EventMetadata metadata = this.CreateEventMetadata(errno);
+                    this.Tracer.RelatedWarning(
+                        metadata,
+                        $"{nameof(MacFileBasedLock)}.{nameof(this.TryAcquireLock)}: Failed to write signature");
+                }
+            }
+            else
+            {
+                int errno = Marshal.GetLastWin32Error();
+                EventMetadata metadata = this.CreateEventMetadata(errno);
+                this.Tracer.RelatedWarning(
+                    metadata,
+                    $"{nameof(MacFileBasedLock)}.{nameof(this.TryAcquireLock)}: Failed to truncate lock file");
+            }
+
+            return true;
+        }
+
+        public override void Dispose()
+        {
+            if (this.lockFileDescriptor != InvalidFileDescriptor)
+            {
+                if (NativeMethods.Close(this.lockFileDescriptor) != 0)
+                {
+                    int errno = Marshal.GetLastWin32Error();

```

Comment:
I would suggest leaving a comment here that an error from close is advisory only

Reply:
:+1: